### PR TITLE
guix: make git tag and commit known prior to building

### LINKED ISSUE #22623: [22.0] build: Version info ignores a tag
When building against a tagged commit (e.g., https://github.com/bitcoin/bitcoin/releases/tag/v22.0rc2) the version info should pull a tag rather a git commit hash.

Compare:
```
$ bin/bitcoind -version
Bitcoin Core version v0.21.0rc2
Copyright (C) 2009-2020 The Bitcoin Core developers

Please contribute if you find Bitcoin Core useful. Visit
<https://bitcoincore.org/> for further information about the software.
The source code is available from <https://github.com/bitcoin/bitcoin>.

This is experimental software.
Distributed under the MIT software license, see the accompanying file COPYING
or <https://opensource.org/licenses/MIT>

```

and

```
$ bin/bitcoind -version
Bitcoin Core version v22.0.0-g873fbc745d037ad43570f81e58334c397bc95477
```

### PR DESCRIPTION
**Pull request by achow101**
Guix currently builds from the source archive produced by `git archive` in a location outside of the git tree. This means that `share/genbuild.sh` will be unable to determine the version number from git because there is no git information available to it during the build. In order for the correct version to be set, `share/genbuild.sh` will now fallback to `GIT_TAG` and `GIT_COMMIT` environment variables if they are set. Guix makes use of this by setting `GIT_TAG` and `GIT_COMMIT` before leaving the git tree so that during the build, `share/genbuild.sh` will create a `build.h` file containing the correct values for the version string.

Fixes #22623

### CODE REVIEW

**Comment by achow101** [comment]
Needs 22.0 backport

**Comment by fanquake** [comment]
```bash
In contrib/guix/libexec/build.sh line 243:
export GIT_TAG=$(git describe --exact-match 2> /dev/null)
       ^-----^ SC2155: Declare and assign separately to avoid masking return values.


In contrib/guix/libexec/build.sh line 244:
export GIT_COMMIT=$(git rev-parse --short=12 HEAD)
       ^--------^ SC2155: Declare and assign separately to avoid masking return values.

For more information:
  https://www.shellcheck.net/wiki/SC2155 -- Declare and assign separately to ...
^---- failure generated from test/lint/lint-shell.sh
```

File: share/genbuild.sh [hunk 1]
```diff
@@ -18,17 +18,14 @@ else
     exit 1
 fi
 
-GIT_TAG=""
-GIT_COMMIT=""
+GIT_TAG="${GIT_TAG:-}"
+GIT_COMMIT="${GIT_COMMIT:-}"
```
Comment (hebasto) [inline share/genbuild.sh:22]: 6b6b8e7edeba3360738fb40ebc12872a81089967

Should we check `GIT_TAG` and `GIT_COMMIT` for emptiness, and run the following `if` block when they are empty?

File: share/genbuild.sh [hunk 2]
```diff
@@ -18,17 +18,14 @@ else
...
 if [ "${BITCOIN_GENBUILD_NO_GIT}" != "1" ] && [ -e "$(command -v git)" ] && [ "$(git rev-parse --is-inside-work-tree 2>/dev/null)" = "true" ]; then
     # clean 'dirty' status of touched files that haven't been modified
     git diff >/dev/null 2>/dev/null
 
     # if latest commit is tagged and not dirty, then override using the tag name
-    RAWDESC=$(git describe --abbrev=0 2>/dev/null)
-    if [ "$(git rev-parse HEAD)" = "$(git rev-list -1 $RAWDESC 2>/dev/null)" ]; then
-        git diff-index --quiet HEAD -- && GIT_TAG=$RAWDESC
-    fi
+    git diff-index --quiet HEAD -- && GIT_TAG=$(git describe --exact-match 2>/dev/null)
```
Comment (hebasto) [inline share/genbuild.sh:28]: 6b6b8e7edeba3360738fb40ebc12872a81089967

Why this change is required?

File: contrib/guix/libexec/build.sh [hunk 3]
```diff
@@ -236,6 +236,16 @@ mkdir -p "$OUTDIR"
...
 
+# The build will use the extracted git archive in a build directory not under the git work tree.
+# This will result in genbuild.sh being unable to determine the actual git tag for the version
+# string reported by the binary. By exporting GIT_TAG and GIT_COMMIT now, genbuild.sh will know
+# what tag and commit to use.
+# We want "git describe" to not cause an exit on failure and just be set to ""
+# shellcheck disable=SC2155
+export GIT_TAG=$(git describe --exact-match 2> /dev/null)
+git_commit=$(git rev-parse --short=12 HEAD)
+export GIT_COMMIT=${git_commit}
```
Comment (hebasto) [inline contrib/guix/libexec/build.sh:247]: cb4b1906bfbd636df8e7dd7fedee5d0951d2e3b1

Why not just
```suggestion
export GIT_COMMIT=$(git rev-parse --short=12 HEAD)
```
?

**Review by hebasto — COMMENTED** [review]
Concept ACK.

Btw, a top commit hash is [available](https://git-scm.com/docs/git-archive) in git archive:
> Additionally the commit ID is stored in a global extended pax header if the tar format is used; it can be extracted using `git get-tar-commit-id`.

File: share/genbuild.sh (hunk 1 above)
Comment (achow101) [reply share/genbuild.sh:22]: No, I did it this way specifically to allow for `GIT_TAG` and `GIT_COMMIT` to be fallbacks. So if we are in the git work tree, then it won't be possible to accidentally override them with the incorrect values.

File: share/genbuild.sh (hunk 2 above)
Comment (achow101) [reply share/genbuild.sh:28]: It isn't, but it's simpler.

File: contrib/guix/libexec/build.sh (hunk 3 above)
Comment (achow101) [reply contrib/guix/libexec/build.sh:247]: shellcheck complained. https://github.com/koalaman/shellcheck/wiki/SC2155

Comment (hebasto) [reply contrib/guix/libexec/build.sh:247]:
```suggestion
GIT_COMMIT=$(git rev-parse --short=12 HEAD)
export GIT_COMMIT
```

to do not introduce a new variable `git_commit`?

**Comment by achow101** [comment]
Closing in favor of #22685. I think using the version already being set in source is better than trying to extract it from git tags.
